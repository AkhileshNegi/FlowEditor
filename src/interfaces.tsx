import {FlowLoaderComp} from './components/FlowLoaderComp'
import {NodeComp} from './components/NodeComp'
import {FlowMutator} from './components/FlowMutator';
import {NodeFormComp} from './components/NodeFormComp';

export interface FlowContext {
    flow: FlowLoaderComp;
    node?: NodeComp;
}

export interface ContactField {
    uuid: string;
    name: string;
}

export interface Group {
    uuid: string;
    name: string;
}

export interface UINode {
    position: LocationProps;
}

export interface UIMetaDataProps {
    nodes: {[key: string]: UINode};
}

export interface LocationProps {
    x: number;
    y: number;
}


export interface NodeEditorState {

}

export interface NodeEditorProps {
    type: string;
    uuid: string;
    
    mutator?: FlowMutator;

    // creating a new node from dragging
    draggedFrom?: DragPoint;

    // node to add a new action to 
    addToNode?: string;

    // location to create new nodes
    newPosition?: LocationProps;

    // exits
    exits?: ExitProps[]
}

export interface ActionProps extends NodeEditorProps {
    dragging?: boolean;
}

export interface AddToGroupProps extends ActionProps {
    name: string;
    group: string;
}

export interface SaveToContactProps extends ActionProps {
    name: string;
    value: string;
    field: string;
}

export interface SendMessageProps extends ActionProps {
    text: string;
}

export interface SetLanguageProps extends ActionProps {
    language: string;
}

export interface WebhookProps extends ActionProps {
    url: string;
    method: string;
}

export interface RouterProps extends NodeEditorProps {
    type: string;
}

export interface CaseProps {
    uuid: string;
    type: string;
    arguments?: string[];
    exit?: string;

    exitProps?: ExitProps;
    onChanged?: Function;
    moveCase?: Function;
}

export interface SwitchRouterProps extends RouterProps {
    cases: CaseProps[];
    operand: string;
    default: string;

    exits?: ExitProps[];
}

export interface SwitchRouterState extends NodeEditorState {

}

export interface RandomRouterProps extends RouterProps {

}

export interface ExitProps {
    uuid: string;
    destination: string;
    name: string;
}

export interface NodeProps {
    uuid: string;
    exits: ExitProps[];
    
    router?: any;
    actions?: ActionProps[];
    wait?: any;
    _ui?: UINode;

    // hook for updating the flow
    mutator?: FlowMutator;

    // a ghost node dragged from somewhere
    draggedFrom?: DragPoint

    // a connection that needs to be wired on mounting
    pendingConnection?: DragPoint;

}

/**
 * A point in the flow from which a drag is initiated
 */
export interface DragPoint {
    exitUUID: string;
    nodeUUID: string;
    onResolved?: Function;
}

export interface TypeConfig {
    type: string;
    name: string;
    description: string;
    
    form?: { new (props: NodeEditorProps): any};
    component?: { new (props: NodeEditorProps): any};
}

export interface Operator {
    type: string;
    name: string;
    verboseName: string;
    operands: number;
}

export interface FlowDefinition {
    localization:  {[lang: string]: {[uuid: string]: any}};
    nodes: NodeProps[]
    uuid: string
    _ui: UIMetaDataProps
    
}

export interface SearchResult {
    name: string,
    id: string,
    type: string,
    prefix?: string,
    extraResult?: boolean
}

export interface ContactFieldResult extends SearchResult {
    key?: string
}