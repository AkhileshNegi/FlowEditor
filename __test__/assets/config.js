const config = {
    flow: 'boring'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  contacts: 'contacts',
                  fields: 'fields',
                  labels: 'labels',
                  activity: '',
                  engine: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows',
                  groups: '/assets/groups',
                  contacts: '/assets/contacts',
                  fields: '/assets/fields',
                  labels: '/assets/labels',
                  activity: '',
                  engine: ''
              }
          });
