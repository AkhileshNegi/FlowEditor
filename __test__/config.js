const config = {
    flow: 'a4f64f1b-85bc-477e-b706-de313a022979'
};

module.exports =
    process.env.NODE_ENV === 'production'
        ? Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: 'flows',
                  groups: 'groups',
                  recipients: 'recipients',
                  fields: 'fields',
                  labels: 'labels',
                  languages: 'languages',
                  channels: 'channels',
                  environment: 'environment',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          })
        : Object.assign({}, config, {
              localStorage: true,
              endpoints: {
                  flows: '/assets/flows.json',
                  groups: '/assets/groups.json',
                  fields: '/assets/fields.json',
                  recipients: '/assets/recipients.json',
                  labels: '/assets/labels.json',
                  languages: '/assets/languages.json',
                  channels: '/assets/channels.json',
                  environment: '/assets/environment.json',
                  activity: '',
                  simulateStart: '',
                  simulateResume: ''
              }
          });
