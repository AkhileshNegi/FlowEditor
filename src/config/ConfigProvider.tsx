import * as React from 'react';
import { Endpoints, FlowEditorConfig } from '~/flowTypes';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
    config: FlowEditorConfig;
    children: any;
}

export interface ConfigProviderContext {
    endpoints: Endpoints;
    flow: string;
    debug: boolean;
    showDownload: boolean;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
    'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
    public static childContextTypes = {
        assetService: fakePropType,
        endpoints: fakePropType,
        flow: fakePropType,
        debug: fakePropType,
        showDownload: fakePropType
    };

    constructor(props: ConfigProviderProps) {
        super(props);

        if (React.Children.count(props.children) > 1) {
            throw new Error(SINGLE_CHILD_ERROR);
        } else if (!React.isValidElement(props.children)) {
            throw new Error(VALID_CHILD_ERROR);
        }
    }

    public getChildContext(): ConfigProviderContext {
        return {
            endpoints: this.props.config.endpoints,
            flow: this.props.config.flow,
            debug: this.props.config.debug,
            showDownload: this.props.config.showDownload
        };
    }

    public render(): JSX.Element {
        return React.Children.only(this.props.children);
    }
}
