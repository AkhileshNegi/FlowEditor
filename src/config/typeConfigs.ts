import AddLabelsComp from '../component/actions/AddLabels/AddLabels';
import AddLabelsForm from '../component/actions/AddLabels/AddLabelsForm';
import CallWebhookComp from '../component/actions/CallWebhook/CallWebhook';
import AddGroupsForm from '../component/actions/ChangeGroups/AddGroupsForm';
import ChangeGroupsComp from '../component/actions/ChangeGroups/ChangeGroups';
import RemoveGroupsForm from '../component/actions/ChangeGroups/RemoveGroupsForm';
import MissingComp from '../component/actions/Missing/Missing';
import SendBroadcastComp from '../component/actions/SendBroadcast/SendBroadcast';
import SendBroadcastForm from '../component/actions/SendBroadcast/SendBroadcastForm';
import { SendBroadcastFormHelper } from '../component/actions/SendBroadcast/SendBroadcastFormHelper';
import SendEmailComp from '../component/actions/SendEmail/SendEmail';
import SendEmailForm from '../component/actions/SendEmail/SendEmailForm';
import SendMsgComp from '../component/actions/SendMsg/SendMsg';
import SendMsgForm from '../component/actions/SendMsg/SendMsgForm';
import SetContactAttrib from '../component/actions/SetContactAttrib/SetContactAttrib';
import SetContactAttribForm from '../component/actions/SetContactAttrib/SetContactAttribForm';
import { SetContactAttribFormHelper } from '../component/actions/SetContactAttrib/SetContactAttribFormHelper';
import SetRunResultComp from '../component/actions/SetRunResult/SetRunResult';
import SetRunResultForm from '../component/actions/SetRunResult/SetRunResultForm';
import StartFlowComp from '../component/actions/StartFlow/StartFlow';
import StartSessionComp from '../component/actions/StartSession/StartSession';
import StartSessionForm from '../component/actions/StartSession/StartSessionForm';
import { StartSessionFormHelper } from '../component/actions/StartSession/StartSessionFormHelper';
import GroupsRouter from '../component/routers/GroupsRouter';
import SubflowRouter from '../component/routers/SubflowRouter';
import SwitchRouter from '../component/routers/SwitchRouter';
import WebhookRouter from '../component/routers/WebhookRouter';
import { AnyAction, RouterTypes, UINodeTypes } from '../flowTypes';
import { NodeEditorForm } from '../store/nodeEditor';

/*
Old name	                New name	                Event(s) generated

add_urn	                    add_contact_urn	            contact_urn_added
add_to_group	            add_contact_groups	        contact_groups_added
remove_from_group	        remove_contact_groups	    contact_groups_removed
set_preferred_channel	    set_contact_channel	        contact_channel_changed
update_contact	            set_contact_name	        contact_name_changed
update_contact	            set_contact_language	    contact_language_changed
update_contact	            set_contact_timezone	    contact_timezone_changed
save_contact_field      	set_contact_field	        contact_field_changed
save_flow_result	        set_run_result	            run_result_changed
call_webhook	            call_webhook	            webhook_called
add_label	                add_input_labels	        input_labels_added
reply	                    send_msg	                msg_created
send_email	                send_email	                email_created / email_sent
send_msg	                send_broadcast	            broadcast_created
start_flow	                start_flow	                flow_triggered
start_session	            start_session	            session_triggered
*/

export const enum Types {
    add_contact_urn = 'add_contact_urn',
    add_contact_groups = 'add_contact_groups',
    add_input_labels = 'add_input_labels',
    remove_contact_groups = 'remove_contact_groups',
    set_contact_channel = 'set_contact_channel',
    set_contact_name = 'set_contact_name',
    set_contact_field = 'set_contact_field',
    set_run_result = 'set_run_result',
    call_webhook = 'call_webhook',
    send_msg = 'send_msg',
    send_email = 'send_email',
    send_broadcast = 'send_broadcast',
    start_flow = 'start_flow',
    start_session = 'start_session',
    split_by_expression = 'split_by_expression',
    split_by_groups = 'split_by_groups',
    wait_for_response = 'wait_for_response',

    missing = 'missing'
}

export enum Mode {
    EDITING = 0x1,
    TRANSLATING = 0x2,
    ALL = EDITING | TRANSLATING
}

export interface FormHelper {
    actionToState: (action: AnyAction, actionType?: Types) => NodeEditorForm;
    stateToAction: (actionUUID: string, formState: NodeEditorForm, formType?: Types) => AnyAction;
}

export interface Type {
    type: Types;
    name: string;
    description: string;
    allows(mode: Mode): boolean;
    component?: React.SFC<AnyAction>;
    form?: React.ComponentClass<any>;
    formHelper?: FormHelper;
    advanced?: Mode;
    aliases?: string[];
}

export interface TypeMap {
    [propName: string]: Type;
}

export type GetTypeConfig = (type: string) => Type;

export function allows(mode: Mode): boolean {
    return (this.advanced & mode) === mode;
}

const ContactAttribHelper = new SetContactAttribFormHelper();

export const typeConfigList: Type[] = [
    /** Actions */
    {
        type: Types.missing,
        name: 'Missing',
        description: ' ** Unsupported ** ',
        component: MissingComp,
        allows
    },
    {
        type: Types.send_msg,
        name: 'Send Message',
        description: 'Send the contact a message',
        form: SendMsgForm,
        component: SendMsgComp,
        advanced: Mode.EDITING,
        allows
    },
    {
        type: Types.send_broadcast,
        name: 'Send Broadcast',
        description: 'Send somebody else a message',
        form: SendBroadcastForm,
        formHelper: new SendBroadcastFormHelper(),
        component: SendBroadcastComp,
        allows
    },
    {
        type: Types.add_input_labels,
        name: 'Add Labels',
        description: 'Label the incoming message',
        form: AddLabelsForm,
        component: AddLabelsComp,
        allows
    },
    {
        type: Types.add_contact_groups,
        name: 'Add to Group',
        description: 'Add the contact to a group',
        form: AddGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.remove_contact_groups,
        name: 'Remove from Group',
        description: 'Remove the contact from a group',
        form: RemoveGroupsForm,
        component: ChangeGroupsComp,
        allows
    },
    {
        type: Types.set_contact_field,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.set_contact_name,
        name: 'Update Contact',
        description: 'Update the contact',
        form: SetContactAttribForm,
        formHelper: ContactAttribHelper,
        component: SetContactAttrib,
        allows
    },
    {
        type: Types.send_email,
        name: 'Send Email',
        description: 'Send an email',
        form: SendEmailForm,
        component: SendEmailComp,
        allows
    },
    {
        type: Types.set_run_result,
        name: 'Save Flow Result',
        description: 'Save a result for this flow',
        form: SetRunResultForm,
        component: SetRunResultComp,
        allows
    },
    // {type: 'set_preferred_channel', name: 'Set Preferred Channel', description: 'Set their preferred channel', component: Missing},
    /** Hybrids */
    {
        type: Types.call_webhook,
        name: 'Call Webhook',
        description: 'Call a webook',
        form: WebhookRouter,
        component: CallWebhookComp,
        advanced: Mode.EDITING,
        aliases: [UINodeTypes.webhook],
        allows
    },
    {
        type: Types.start_flow,
        name: 'Start a Flow',
        description: 'Enter another flow',
        form: SubflowRouter,
        component: StartFlowComp,
        aliases: [UINodeTypes.subflow],
        allows
    },
    {
        type: Types.start_session,
        name: 'Start Somebody Else',
        description: 'Start somebody else in a flow',
        form: StartSessionForm,
        formHelper: new StartSessionFormHelper(),
        component: StartSessionComp,
        allows
    },

    /** Routers */
    {
        type: Types.split_by_expression,
        name: 'Split by Expression',
        description: 'Split by a custom expression',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        allows
    },
    {
        type: Types.split_by_groups,
        name: 'Split by Group Membership',
        description: 'Split by group membership',
        form: GroupsRouter,
        allows
    },
    {
        type: Types.wait_for_response,
        name: 'Wait for Response',
        description: 'Wait for the contact to respond',
        form: SwitchRouter,
        advanced: Mode.TRANSLATING,
        aliases: [RouterTypes.switch],
        allows
    }
    // {type: 'random', name: 'Random Split', description: 'Split them up randomly', form: RandomRouterForm}
];

export const typeConfigMap: TypeMap = typeConfigList.reduce((map: TypeMap, typeConfig: Type) => {
    map[typeConfig.type] = typeConfig;
    if (typeConfig.aliases) {
        typeConfig.aliases.forEach((alias: string) => (map[alias] = typeConfig));
    }
    return map;
}, {});

/**
 * Shortcut for constant lookup of type config in type configs map
 * @param {string} type - The type of the type config to return, e.g. 'send_msg'
 * @returns {Object} - The type config found at typeConfigs[type] or -1
 */
export const getTypeConfig = (type: Types | RouterTypes | UINodeTypes): Type => {
    let actionConfig = typeConfigMap[type];

    if (!actionConfig) {
        actionConfig = typeConfigMap.missing;
    }
    return actionConfig;
};
