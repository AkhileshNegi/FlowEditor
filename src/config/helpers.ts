import { Operator, Type } from '~/config';
import { FlowTypes, FlowTypeVisibility } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    // return filterVisibility(operators, flowType);
    return operators;
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
