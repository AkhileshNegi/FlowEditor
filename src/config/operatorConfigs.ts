import { HIDDEN, Operator, OperatorMap, Operators, TEXT_TYPES } from '~/config/interfaces';

export const operatorConfigList: Operator[] = [
    {
        type: Operators.has_any_word,
        verboseName: 'has any of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_all_words,
        verboseName: 'has all of the words',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_phrase,
        verboseName: 'has the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_only_phrase,
        verboseName: 'has only the phrase',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_beginning,
        verboseName: 'starts with',
        operands: 1
    },
    {
        type: Operators.has_text,
        verboseName: 'has some text',
        operands: 0,
        categoryName: 'Has Text',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_number,
        verboseName: 'has a number',
        operands: 0,
        categoryName: 'Has Number'
    },
    {
        type: Operators.has_number_between,
        verboseName: 'has a number between',
        operands: 2
    },
    {
        type: Operators.has_number_lt,
        verboseName: 'has a number below',
        operands: 1
    },
    {
        type: Operators.has_number_lte,
        verboseName: 'has a number at or below',
        operands: 1
    },
    {
        type: Operators.has_number_eq,
        verboseName: 'has a number equal to',
        operands: 1
    },
    {
        type: Operators.has_number_gte,
        verboseName: 'has a number at or above',
        operands: 1
    },
    {
        type: Operators.has_number_gt,
        verboseName: 'has a number above',
        operands: 1
    },
    {
        type: Operators.has_date,
        verboseName: 'has a date',
        operands: 0,
        categoryName: 'Has Date',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_lt,
        verboseName: 'has a date before',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_eq,
        verboseName: 'has a date equal to',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_date_gt,
        verboseName: 'has a date after',
        operands: 1,
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_run_status,
        verboseName: 'has a run status of',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_group,
        verboseName: 'is in the group',
        operands: 1,
        visibility: HIDDEN
    },
    {
        type: Operators.has_phone,
        verboseName: 'has a phone number',
        operands: 0,
        categoryName: 'Has Phone'
    },
    {
        type: Operators.has_email,
        verboseName: 'has an email',
        operands: 0,
        categoryName: 'Has Email',
        visibility: TEXT_TYPES
    },
    {
        type: Operators.has_error,
        verboseName: 'has an error',
        operands: 0,
        categoryName: 'Has Error',
        visibility: HIDDEN
    },
    {
        type: Operators.has_value,
        verboseName: 'is not empty',
        operands: 0,
        categoryName: 'Not Empty'
    },
    {
        type: Operators.has_wait_timed_out,
        verboseName: null,
        operands: 0,
        visibility: HIDDEN
    },
    {
        type: Operators.has_pattern,
        verboseName: 'matches regex',
        operands: 1
    }
];

export const operatorConfigMap: OperatorMap = operatorConfigList.reduce(
    (map: OperatorMap, operatorConfig: Operator) => {
        map[operatorConfig.type] = operatorConfig;
        return map;
    },
    {}
);

/**
 * Shortcut for constant lookup of operator config in operator configs map
 * @param {string} type - The type of the operator config to return, e.g. 'send_msg'
 * @returns {Object} - The operator config found at operatorConfigs[type] or -1
 */
export const getOperatorConfig = (type: Operators): Operator => operatorConfigMap[type];
