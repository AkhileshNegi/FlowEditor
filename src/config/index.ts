import ConfigProvider, { ConfigProviderContext } from './ConfigProvider';
import { Type, Mode, typeConfigList, getTypeConfig } from './typeConfigs';
import { Operator, operatorConfigList, getOperatorConfig } from './operatorConfigs';

export {
    ConfigProviderContext,
    Type,
    Mode,
    Operator,
    typeConfigList,
    getTypeConfig,
    operatorConfigList,
    getOperatorConfig
};

export default ConfigProvider;
