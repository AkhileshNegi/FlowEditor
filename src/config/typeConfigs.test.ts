import { Types } from "config/interfaces";
import {
  getTypeConfig,
  typeConfigList,
  typeConfigMap
} from "config/typeConfigs";

describe("typeConfigs", () => {
  it("should provide type configs", () => {
    expect(typeConfigList).toMatchSnapshot();
    expect(typeConfigMap).toMatchSnapshot();

    Object.keys(typeConfigMap).forEach((key: Types) =>
      expect(getTypeConfig(key)).toEqual(typeConfigMap[key])
    );
  });
});
