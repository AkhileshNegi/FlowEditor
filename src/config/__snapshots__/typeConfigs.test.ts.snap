// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeConfigs should provide type configs 1`] = `
Array [
  Object {
    "component": [Function],
    "description": " ** Unsupported ** ",
    "name": "Missing",
    "type": "missing",
  },
  Object {
    "advanced": 3,
    "component": [Function],
    "description": "Send the contact a message",
    "form": [Function],
    "formHelper": SendMsgFormHelper {},
    "name": "Send Message",
    "type": "send_msg",
  },
  Object {
    "component": [Function],
    "description": "Send somebody else a message",
    "form": [Function],
    "formHelper": SendBroadcastFormHelper {},
    "name": "Send Broadcast",
    "type": "send_broadcast",
  },
  Object {
    "component": [Function],
    "description": "Label the incoming message",
    "form": [Function],
    "formHelper": AddLabelsFormHelper {},
    "name": "Add Labels",
    "type": "add_input_labels",
  },
  Object {
    "component": [Function],
    "description": "Add the contact to a group",
    "form": [Function],
    "formHelper": AddGroupsFormHelper {},
    "name": "Add to Group",
    "type": "add_contact_groups",
  },
  Object {
    "component": [Function],
    "description": "Remove the contact from a group",
    "form": [Function],
    "formHelper": RemoveGroupsFormHelper {},
    "name": "Remove from Group",
    "type": "remove_contact_groups",
  },
  Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_field",
  },
  Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_name",
  },
  Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_language",
  },
  Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_channel",
  },
  Object {
    "component": [Function],
    "description": "Send an email",
    "form": [Function],
    "formHelper": SendEmailFormHelper {},
    "name": "Send Email",
    "type": "send_email",
  },
  Object {
    "component": [Function],
    "description": "Save a result for this flow",
    "form": [Function],
    "formHelper": SetRunResultFormHelper {},
    "name": "Save Flow Result",
    "type": "set_run_result",
  },
  Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
  Object {
    "aliases": Array [
      "subflow",
    ],
    "component": [Function],
    "description": "Enter another flow",
    "form": [Function],
    "name": "Start a Flow",
    "type": "start_flow",
  },
  Object {
    "component": [Function],
    "description": "Start somebody else in a flow",
    "form": [Function],
    "formHelper": StartSessionFormHelper {},
    "name": "Start Somebody Else",
    "type": "start_session",
  },
  Object {
    "advanced": 2,
    "description": "Split by a custom expression",
    "form": [Function],
    "name": "Split by Expression",
    "type": "split_by_expression",
  },
  Object {
    "description": "Split by group membership",
    "form": [Function],
    "name": "Split by Group Membership",
    "type": "split_by_groups",
  },
  Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "description": "Wait for the contact to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
]
`;

exports[`typeConfigs should provide type configs 2`] = `
Object {
  "add_contact_groups": Object {
    "component": [Function],
    "description": "Add the contact to a group",
    "form": [Function],
    "formHelper": AddGroupsFormHelper {},
    "name": "Add to Group",
    "type": "add_contact_groups",
  },
  "add_input_labels": Object {
    "component": [Function],
    "description": "Label the incoming message",
    "form": [Function],
    "formHelper": AddLabelsFormHelper {},
    "name": "Add Labels",
    "type": "add_input_labels",
  },
  "call_webhook": Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
  "missing": Object {
    "component": [Function],
    "description": " ** Unsupported ** ",
    "name": "Missing",
    "type": "missing",
  },
  "remove_contact_groups": Object {
    "component": [Function],
    "description": "Remove the contact from a group",
    "form": [Function],
    "formHelper": RemoveGroupsFormHelper {},
    "name": "Remove from Group",
    "type": "remove_contact_groups",
  },
  "send_broadcast": Object {
    "component": [Function],
    "description": "Send somebody else a message",
    "form": [Function],
    "formHelper": SendBroadcastFormHelper {},
    "name": "Send Broadcast",
    "type": "send_broadcast",
  },
  "send_email": Object {
    "component": [Function],
    "description": "Send an email",
    "form": [Function],
    "formHelper": SendEmailFormHelper {},
    "name": "Send Email",
    "type": "send_email",
  },
  "send_msg": Object {
    "advanced": 3,
    "component": [Function],
    "description": "Send the contact a message",
    "form": [Function],
    "formHelper": SendMsgFormHelper {},
    "name": "Send Message",
    "type": "send_msg",
  },
  "set_contact_channel": Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_channel",
  },
  "set_contact_field": Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_field",
  },
  "set_contact_language": Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_language",
  },
  "set_contact_name": Object {
    "component": [Function],
    "description": "Update the contact",
    "form": [Function],
    "formHelper": SetContactAttribFormHelper {},
    "name": "Update Contact",
    "type": "set_contact_name",
  },
  "set_run_result": Object {
    "component": [Function],
    "description": "Save a result for this flow",
    "form": [Function],
    "formHelper": SetRunResultFormHelper {},
    "name": "Save Flow Result",
    "type": "set_run_result",
  },
  "split_by_expression": Object {
    "advanced": 2,
    "description": "Split by a custom expression",
    "form": [Function],
    "name": "Split by Expression",
    "type": "split_by_expression",
  },
  "split_by_groups": Object {
    "description": "Split by group membership",
    "form": [Function],
    "name": "Split by Group Membership",
    "type": "split_by_groups",
  },
  "start_flow": Object {
    "aliases": Array [
      "subflow",
    ],
    "component": [Function],
    "description": "Enter another flow",
    "form": [Function],
    "name": "Start a Flow",
    "type": "start_flow",
  },
  "start_session": Object {
    "component": [Function],
    "description": "Start somebody else in a flow",
    "form": [Function],
    "formHelper": StartSessionFormHelper {},
    "name": "Start Somebody Else",
    "type": "start_session",
  },
  "subflow": Object {
    "aliases": Array [
      "subflow",
    ],
    "component": [Function],
    "description": "Enter another flow",
    "form": [Function],
    "name": "Start a Flow",
    "type": "start_flow",
  },
  "switch": Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "description": "Wait for the contact to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
  "wait_for_response": Object {
    "advanced": 2,
    "aliases": Array [
      "switch",
    ],
    "description": "Wait for the contact to respond",
    "form": [Function],
    "name": "Wait for Response",
    "type": "wait_for_response",
  },
  "webhook": Object {
    "advanced": 1,
    "aliases": Array [
      "webhook",
    ],
    "component": [Function],
    "description": "Call a webook",
    "form": [Function],
    "name": "Call Webhook",
    "type": "call_webhook",
  },
}
`;
